### ISSUE TITLE
substitute-path and bazel problems

### ISSUE THREAD

**Issue by oakad** [go-delve/delve#3333]
Follow up to #245.

In some situations it is imperative to be able to configure the debugger locally to the invocation without relying on remote side configuration.

In various container/remote build environments there's information available to the tool invoking the dlv in headless mode which is not available to the debugging client. It is imperative, therefore, to be able to supply this information to the dlv prior to the connection from the client.

This can be done in the config file, but config file location is stupidly restrictive and can not be reliably overridden in all cases.
This can be done with init file, but in headless mode dlv refuses to accept one.

**Comment by aarzilli** [comment]
AFAIK the only part of the configuration file that would apply to a headless instance is `debug-info-directories`, everything else is discarded. There's a bunch of problems with allowing an init file in headless mode, both technically (can't work without accept-multiclient) and conceptual (the code that executes the init file is part of the client).

We could potentially add a flag to specify the location of the configuration file, although it can be configured with XDG_CONFIG_HOME, but the context of this issue makes me think that we shouldn't since it wouldn't do what you want it to do.

I think you are under the impression that the server applies the substitute path rules, but it does not: it's done client side and each client could have different rules.

**Comment by oakad** [comment]
It does not, I have figured it out the hard way. :-)

However, I do believe it is an important feature for something like Bazel. Bazel is very protective of its source paths; it is possible to figure them out by means of hackish build action, but it's about it. No way to setup the source map for an IDE in a general case.

I have made a rule for invoking a debugger from inside bazel build which builds a substitution map to the tune of `bazel run //my/go/app:with_dlv -- args`. It works pretty well when init file does have an effect, that is, in console.

But to integrate it reasonably with an IDE, the only way is to somehow feed the path substitution info into the "headless mode" dlv, and it appears to be not possible right now.

Of course, there's still going to be some discrepancy between dependency layout expected by the `gopls` and produced by `bazel`, but `rules_go` is evolving toward fixing it.

**Comment by aarzilli** [comment]
I don't understand how this would work. If you are running `bazel run //my/go/app:with_dlv -- args` on your local machine (i.e. the one where you are running delve's client) then you can pass the result to `--init` and there should be no problem. If you are running it on the same machine that has the headless instance then how can it work? How would it know where the source code happens to be on the client machine?

Also if you are talking about IDE integration and gopls you should also consider that DAP has its own way of specifying the substitution map (through launch.json).

**Comment by oakad** [comment]
As I said, it works in console with built in TUI.
Next step was to try to integrate it with VSCode in "headless" mode, which failed miserably.  :-)

> Also if you are talking about IDE integration and gopls you should also consider that DAP has its own way of specifying the substitution map (through launch.json).

That''s exactly what I'm talking about. In bazel's case there's no reliable way to figure the paths ahead of time and write an "always  working" launch.json. By "always working" I mean something that can be put into project documentation and passed on to fellow developers, as opposed to "spend an hour rewriting the substitution map for each additional repo cloned".

**Comment by aarzilli** [comment]
Why not make bazel rewrite launch.json?

**Comment by oakad** [comment]
1. This is the solution I ended up using for now. I made my rule produce "launch.json snippet" which is then merged into the launch.json
2. It still requires teaching users things about what it's doing, cannot be committed into common repo (because it ends up different for different users) and so on.
3. If headless dlv could rewrite paths on its side, it will be a single, fixed launch.json entry, perfectly committable.
